Add new Tier 3 targets for ARMv6

thejpster opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

Adds three new targets to support ARMv6 processors running bare-metal:

* `armv6-none-eabi` - Arm ISA, soft-float
* `armv6-none-eabihf` - Arm ISA, hard-float
* `thumbv6-none-eabi` - Thumb-1 ISA, soft-float

There is no `thumbv6-none-eabihf` target because as far as I can tell, hard-float isn't support with the Thumb-1 instruction set (and you need the ARMv6T2 extension to enable Thumb-2 support).

The targets require ARMv6K as a minimum, which allows the two Arm ISA targets to have full CAS atomics. LLVM has a bug which means it emits some ARMv6K instructions even if you only call for ARMv6, and as no-one else has noticed the bug, and because basically all ARMv6 processors have ARMv6K, I think this is fine. The Thumb target also doesn't have any kind of atomics, just like the Armv5TE and Armv4 targets, because LLVM was emitting library calls to emulate them.

Testing will be added to https://github.com/rust-embedded/aarch32 once the target is accepted. I already have tests for the other non-M arm-none-eabi targets, and those tests pass on these targets.

> A tier 3 target must have a designated developer or developers (the "target maintainers") on record to be CCed when issues arise regarding the target. (The mechanism to track and CC such developers may evolve over time.)

I have listed myself. If accepted, I'll talk to the Embedded Devices Working Group about adding this one to the rosta with all the others they support.

> Targets must use naming consistent with any existing targets; for instance, a target for the same CPU or OS as an existing Rust target should use the same name for that CPU or OS. Targets should normally use the same names and naming conventions as used elsewhere in the broader ecosystem beyond Rust (such as in other toolchains), unless they have a very good reason to diverge. Changing the name of a target can be highly disruptive, especially once the target reaches a higher tier, so getting the name right is important even for a tier 3 target. 

You might prefer `arm-none-eabi`, because `arm-unknown-linux-gnu` is an ARMv6 target - the implicit rule seems to be that if the Arm architecture version isn't specified, it's assumed to be v6. However, `armv6-none-eabi` seemed to fit better between `armv5te-none-eabi` and `armv7a/armv7r-none-eabi`.

The hamming distance between `thumbv6-none-eabi` and `thumbv6m-none-eabi` is unfortunately low, but I don't know how to make it better. They *are* the ARMv6 and ARMv6-M targets, and its perhaps not worse than `armv7a-none-eabi` and `armv7r-none-eabi`.

> Tier 3 targets may have unusual requirements to build or use, but must not create legal issues or impose onerous legal terms for the Rust project or for Rust developers or users. 

No different to any other arm-none-eabi target.

> Neither this policy nor any decisions made regarding targets shall create any binding agreement or estoppel by any party. If any member of an approving Rust team serves as one of the maintainers of a target, or has any legal or employment requirement (explicit or implicit) that might affect their decisions regarding a target, they must recuse themselves from any approval decisions regarding the target's tier status, though they may otherwise participate in discussions. 

Noted.

> Tier 3 targets should attempt to implement as much of the standard libraries as possible and appropriate...

Same as other arm-none-eabi targets.

> The target must provide documentation for the Rust community explaining how to build for the target, using cross-compilation if possible.

Same as other arm-none-eabi targets.

> Tier 3 targets must not impose burden on the authors of pull requests, or other developers in the community, to maintain the target. In particular, do not post comments (automated or manual) on a PR that derail or suggest a block on the PR based on a tier 3 target. Do not send automated messages or notifications (via any medium, including via @) to a PR author or others involved with a PR regarding a tier 3 target, unless they have opted into such messages. 

Noted.

> Patches adding or updating tier 3 targets must not break any existing tier 2 or tier 1 target, and must not knowingly break another tier 3 target without approval of either the compiler team or the maintainers of the other tier 3 target. 

Noted

> Tier 3 targets must be able to produce assembly using at least one of rustc's supported backends from any host target. (Having support in a fork of the backend is not sufficient, it must be upstream.)

Noted

thejpster committed:
Add ARMv6 bare-metal targets

Three targets, covering A32 and T32 instructions, and soft-float and
hard-float ABIs. Hard-float not available in Thumb mode. Atomics
in Thumb mode require __sync* functions from compiler-builtins.

thejpster committed:
Fix typo in thumbv4t/v5te README

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/20351624965/job/58477476251) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/58477476251) [(plain)](https://github.com/rust-lang/rust/commit/696697d7adbcf69834bdbb4bbf670d19e79e8aa6/checks/58477476251/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[endgroup]
[TIMING:end] tool::ToolBuild { build_compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu, tool: "tidy", path: "src/tools/tidy", mode: ToolBootstrap, source_type: InTree, extra_features: [], allow_features: "", cargo_args: [], artifact_kind: Binary } -- 13.169
[TIMING:end] tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
Diff in /checkout/compiler/rustc_target/src/spec/targets/thumbv6_none_eabi.rs:21:
             features: "+soft-float,+strict-align,+v6k".into(),
             // atomics not available until ARMv6T2
             atomic_cas: false,
-            max_atomic_width: Some(0),           
+            max_atomic_width: Some(0),
             has_thumb_interworking: true,
             ..base::arm_none::opts()
         },
fmt: checked 6609 files
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Build completed unsuccessfully in 0:00:54
  local time: Thu Dec 18 21:22:56 UTC 2025
```

</details>

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/20351894077/job/58478368971) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/58478368971) [(plain)](https://github.com/rust-lang/rust/commit/32cd8e5c29f69d493b6a6f1bcd3bd93bb954842f/checks/58478368971/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[endgroup]
[TIMING:end] tool::ToolBuild { build_compiler: Compiler { stage: 0, host: aarch64-unknown-linux-gnu, forced_compiler: false }, target: aarch64-unknown-linux-gnu, tool: "linkchecker", path: "src/tools/linkchecker", mode: ToolBootstrap, source_type: InTree, extra_features: [], allow_features: "", cargo_args: [], artifact_kind: Binary } -- 0.146
[TIMING:end] tool::Linkchecker { compiler: Compiler { stage: 0, host: aarch64-unknown-linux-gnu, forced_compiler: false }, target: aarch64-unknown-linux-gnu } -- 0.000
##[group]Testing stage1 Linkcheck (aarch64-unknown-linux-gnu)
rustc/print.html:15622: broken intra-doc link - [<code>AtomicU8</code>] and [<code>AtomicI8</code>]
rustc/platform-support/armv6-none-eabi.html:195: broken intra-doc link - [<code>AtomicU8</code>] and [<code>AtomicI8</code>]
checked links in: 22.9s
number of HTML files scanned: 46427
number of HTML redirects found: 14548
number of links checked: 2563503
number of links ignored due to external: 123266
---
Command `/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools-bin/linkchecker /checkout/obj/build/aarch64-unknown-linux-gnu/doc` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1612:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:190:77

Command has failed. Rerun with -v to see more details.
Bootstrap failed while executing `--stage 2 test --skip tests --skip coverage-map --skip coverage-run --skip library --skip tidyselftest`
Build completed unsuccessfully in 0:39:06
  local time: Thu Dec 18 22:12:45 UTC 2025
  network time: Thu, 18 Dec 2025 22:12:46 GMT
##[error]Process completed with exit code 1.
```

</details>

madsmtm commented on library/stdarch/crates/core_arch/src/arm_shared/hints.rs:
```diff
@@ -83,8 +83,12 @@ pub unsafe fn __sevl() {
+    all(target_feature = "v6k", not(target_feature = "thumb-mode")),
+    target_feature = "v6t2",
+    all(target_feature = "v6", target_feature = "mclass"),
+    target_feature = "v7",
```
Q: I believe that it is, but where would I go to verify that the `cfg` here is correct?

thejpster replied:
Added a link to the instruction encoding page in the Arm Armv7 A.R.M.

thejpster committed:
Revised yield hints

Turns out v7 targets always have v6t2 set, so that line was redundant. Also add a link to the Arm Armv7 A.R.M.

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/20368040048/job/58527761421) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/58527761421) [(plain)](https://github.com/rust-lang/rust/commit/4c032134ac613ca32924aa16dd750279bcb9c667/checks/58527761421/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling tier-check v0.1.0 (/checkout/src/tools/tier-check)
[RUSTC-TIMING] tier_check test:false 0.343
    Finished `release` profile [optimized] target(s) in 0.49s
     Running `obj/build/aarch64-unknown-linux-gnu/bootstrap-tools/aarch64-unknown-linux-gnu/release/tier-check /checkout/src/doc/rustc/src/platform-support.md /checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc`
error: target `armv6-none-eabi` is missing from platform-support.md
If this is a new target, please add it to /checkout/src/doc/rustc/src/platform-support.md.
error: target `armv6-none-eabihf` is missing from platform-support.md
If this is a new target, please add it to /checkout/src/doc/rustc/src/platform-support.md.
error: target `thumbv6-none-eabi` is missing from platform-support.md
If this is a new target, please add it to /checkout/src/doc/rustc/src/platform-support.md.
Command `/checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo run --target aarch64-unknown-linux-gnu -Zbinary-dep-depinfo -j 4 -Zroot-dir=/checkout --locked --color=always --release --manifest-path /checkout/src/tools/tier-check/Cargo.toml /checkout/src/doc/rustc/src/platform-support.md /checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc [workdir=/checkout]` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:191:21
Executed at: src/bootstrap/src/core/build_steps/test.rs:3610:55

Command has failed. Rerun with -v to see more details.
Bootstrap failed while executing `--stage 2 test --skip tests --skip coverage-map --skip coverage-run --skip library --skip tidyselftest`
Build completed unsuccessfully in 0:40:04
  local time: Fri Dec 19 11:49:22 UTC 2025
  network time: Fri, 19 Dec 2025 11:49:22 GMT
##[error]Process completed with exit code 1.
```

</details>

taiki-e commented on compiler/rustc_target/src/spec/targets/armv6_none_eabi.rs:
```diff
@@ -0,0 +1,27 @@
+        pointer_width: 32,
+        arch: Arch::Arm,
+        data_layout: "e-m:e-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64".into(),
+        options: TargetOptions {
```
IIRC this target can have `max_atomic_width: Some(64),`. (default is the pointer size)

thejpster commented on src/doc/rustc/src/platform-support/armv6-none-eabi.md:
```diff
@@ -0,0 +1,37 @@
+
+This target assumes your processor has the Armv6K extensions, as basically all
+Armv6 processors do[^1]. The Armv6K extension adds the `LDREXB` and `STREXB`
+instructions required to implement CAS on the [`AtomicU8`] and [`AtomicI8`] types.
```
Added the word 'efficiently' here.

taiki-e commented on compiler/rustc_target/src/spec/targets/thumbv6_none_eabi.rs:
```diff
@@ -0,0 +1,29 @@
+            features: "+soft-float,+strict-align,+v6k".into(),
+            // atomics not available until ARMv6T2
+            atomic_cas: false,
+            max_atomic_width: Some(0),
```
I created a custom target equivalent to this, modified the max_atomic_width, and tried building it, but it failed to compile due to missing `__sync_*` built-ins. (`instruction_set(arm::a32)` on atomic methods also did not work.)

This is not an LLVM error but simply the generation of `__sync_*` builtins, so I believe the correct approach is to implement the `__sync_*` builtins using compiler builtins with instruction_set(arm::a32) and assembly.

taiki-e replied:
> I tried `max_atomic_width: Some(32)` and even with a relaxed load/store I got:
> 
> ```
> rust-lld: error: undefined symbol: __sync_val_compare_and_swap_4
> ```
>
> So for the same reason as ARMv4T and ARMv5TE I propose just switching atomics off here, and letting portable-atomic provide a work-around if it so chooses.

Thanks for confirming. LLVM just generates `__sync_*` builtins instead of raising errors (like they do for __yield), so I believe the situation is simpler.

I believe more appropriate approach here is to implement the `__sync_*` builtins in compiler-builtins with `instruction_set(arm::a32)` and inline assembly.

taiki-e replied:
> I believe more appropriate approach here is to implement the `__sync_*` builtins in compiler-builtins with `instruction_set(arm::a32)` and inline assembly.

Opened https://github.com/rust-lang/compiler-builtins/pull/1050 which implements this approach. 

If you and the compiler-builtins maintainer are okay with that approach, we can set the max-atomic-width to 32 and atomic-cas to true for thumbv6-none-eabi after the new compiler-builtins version containing that change is released.

thejpster commented:
Changes have been proposed to compiler-builtins which rely on this target, so I'm not sure what order we need to merge these PRs.

Maybe merge this where thumbv6 doesn't have CAS, merge the compiler-builtins change, and then update the target to turn CAS on?

@taiki-e would this approach also work for ARMv4T and ARMv5TE (which don't have a DMB equivalent CP15 instruction AFAIK)? Because we turned off all atomics for those recently on the grounds it was completely unclear what the sync* functions would need to do.

davidtwco reviewed (approved):
These targets look good to me. Regarding the naming: `armv6-*` seems fine, we have some `arm-*` targets that are implicitly v6 but we also already have some `armv6-*` targets that are explicitly v6. rust-lang/compiler-builtins#1050 seems like the right approach to me, I'll leave it to @taiki-e to advise on what the right order to merge these is for that.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/150138/a2bc948b7fff08ef1318747e3f4470f54f8a1694..f4826f2e0811a5b1862aadb7459fe63d4aa10c51)*

tgross35 commented:
> Blocked on [rust-lang/compiler-builtins#1050](https://github.com/rust-lang/compiler-builtins/pull/1050), which allows us to support atomics on the thumbv6-none-eabi atomic added here.

I don't think this is blocked, right? CI is passing. Atomics just won't work temporarily which is completely fine for a T3 target.

Happy to merge with David's review if you can rebase so we don't need to wait for a c-b sync.

thejpster commented:
github.com/rust-lang/compiler-builtins/pull/1050 has been merged. I will update this PR to enable 32-bit atomics on these targets, but the target will be broken (LLVM will emit calls to `__sync*` functions that the user almost certainly doesn't have) until compiler-builtins is updated, but that's OK.

wesleywiser reviewed (approved):
r=me with history cleaned up

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/150138/0f145634fc31775477c538c532307213109599ac..1d1133efe8e7cdfbb497d51d641c5d3139e612aa)*
